File: tests/fakedata.py
import random
import string
from datetime import datetime, timedelta

def random_string(length=8):
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))

def random_boolean():
    return random.choice([True, False])

def random_number(min_value=0, max_value=100):
    return random.randint(min_value, max_value)

def random_float(min_value=0.0, max_value=100.0):
    return round(random.uniform(min_value, max_value), 2)

def current_timestamp():
    return int(datetime.now().timestamp())

def random_alert_type():
    return random.choice(["pending", "generated", "resolved"])

def random_audio_source():
    return random.choice(["HDMI", "AUX", "MIC"])

def generate_device_data():
    return {
        "DEVICE_ID": 100001,

        "LOCATION": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "Cell_Info": {
                "lat": random_float(-90, 90),
                "lon": random_float(-180, 180),
            },
            "Installing": random_boolean(),
        },

        "GUEST_REGISTRATION": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "guest_id": random_number(),
            "registering": random_boolean(),
            "guest_age": random_number(18, 80),
            "guest_male": random_boolean(),
        },

        "MEMBER_GUEST_DECLARATION": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "member_keys": [random_boolean() for _ in range(3)],
            "guests": [random_boolean() for _ in range(3)],
            "confidence": random_float(0.0, 1.0),
        },

        "CONFIGURATION": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "differential_mode": random_boolean(),
            "member_keys": [random_boolean() for _ in range(3)],
            "guest_cancellation_time": random_number(),
            "software_version": f"{random_number(1, 10)}.{random_number(0, 9)}.{random_number(0, 9)}",
            "power_pcb_firmware_version": f"{random_number(1, 10)}.{random_number(0, 9)}.{random_number(0, 9)}",
            "remote_firmware_version": f"{random_number(1, 10)}.{random_number(0, 9)}.{random_number(0, 9)}",
            "audio_configuration": [random_boolean() for _ in range(3)],
            "audience_day_start_time": random_number(),
            "no_of_sessions": random_number(1, 10),
        },

        "TAMPER_ALARM": {
            "ID": random_number(),
            "AlertType": random_alert_type(),
            "Type": random_number(),
            "TS": current_timestamp(),
            "Meter_tamper": random_boolean(),
            "Tv_plug_tamper": random_boolean(),
            "Tamper_ts": current_timestamp(),
        },

        "SOS_ALARM": {
            "ID": random_number(),
            "AlertType": random_alert_type(),
            "Type": random_number(),
            "TS": current_timestamp(),
            "sos": random_boolean(),
        },

        "BATTERY_ALARM": {
            "ID": random_number(),
            "AlertType": random_alert_type(),
            "Type": random_number(),
            "TS": current_timestamp(),
            "main_bat_fail": random_boolean(),
            "rtc_fail": random_boolean(),
            "rtc_bat_low": random_boolean(),
        },

        "METER_INSTALLATION": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "HHID": random_string(10),
            "Success": random_boolean(),
        },

        "VOLTAGE_STATS": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "high_rail_voltage": random_float(0, 10),
            "mid_rail_voltage": random_float(0, 10),
            "gsm_rail_voltage": random_float(0, 10),
            "rtc_battery_voltage": random_float(0, 10),
            "li_ion_battery_voltage": random_float(0, 10),
            "remote_battery_voltage": random_float(0, 10),
        },

        "TEMPERATURE_STATS": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "battery_temp": random_float(0, 100),
            "arm_core_temp": random_float(0, 100),
            "power_pcb_temp": random_float(0, 100),
            "rtc_temp": random_float(0, 100),
        },

        "NTP_SYNC": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "server": random_string(15),
            "system_time": current_timestamp(),
            "success": random_boolean(),
            "error_code": random_number(0, 5),
            "drift": random_float(0, 5),
            "jumped": random_boolean(),
        },

        "AUDIENCE_SESSION_CLOSE": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "stop_time": current_timestamp(),
            "viewing_member_keys": [random_boolean() for _ in range(3)],
            "viewing_guests": [random_boolean() for _ in range(3)],
            "tv_on": random_boolean(),
            "last_watermark_id": random_number(),
            "tv_event_ts": current_timestamp(),
            "last_watermark_id_ts": current_timestamp(),
            "marked": random_number(0, 10),
        },

        "NETWORK_LATCH": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "Ip_up": random_boolean(),
            "Sim": random.choice([1, 2]),
        },

        "REMOTE_PAIRING": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "remote_id": random_number(),
            "success": random_boolean(),
        },

        "REMOTE_ACTIVITY": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "lock": random_boolean(),
            "orr": random_boolean(),
            "absent_key_press": random_boolean(),
            "drop": random_boolean(),
        },

        "SIM_ALERT": {
            "ID": random_number(),
            "AlertType": random_alert_type(),
            "Type": random_number(),
            "TS": current_timestamp(),
            "sim1_absent": random_boolean(),
            "sim1_changed": random_boolean(),
            "sim2_absent": random_boolean(),
            "sim2_changed": random_boolean(),
        },

        "SYSTEM_ALARM": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "name": random_string(10),
            "AlertType": random_alert_type(),
            "Details": {"detail_key": random_string(5)},
        },

        "SYSTEM_INFO": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "rpi_serial": random_string(10),
            "pcb_serial": random_string(10),
            "imei": random_string(15),
            "imsi_1": random_string(15),
            "imsi_2": random_string(15),
            "eeprom": random_number(0, 100),
            "wifi_serial": random_string(10),
            "mac_serial": random_string(12),
            "remote_serial": random_number(0, 100),
        },

        "CONFIG_UPDATE": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "key": random_string(10),
            "value": random_string(10),
            "old_value": random_string(10),
        },

        "ALIVE": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "state": random_boolean(),
        },

        "METER_OTA": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "previous": random_string(10),
            "update": random_string(10),
            "success": random_boolean(),
        },

        "BATTERY_VOLTAGE": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "Rtc": random_float(0, 10),
            "Meter": random_float(0, 10),
        },

        "BOOT": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "boot_ts": current_timestamp(),
            "install": random_number(0, 100),
            "last_boot": current_timestamp(),
            "relay_status": [random_boolean() for _ in range(3)],
        },

        "BOOT_V2": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
            "boot_ts": current_timestamp(),
            "install": random_number(0, 100),
            "last_boot": current_timestamp(),
            "last_boot": random_number(0, 100),
        },

          "STB": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
    "derived":  random_boolean(),
  },
  "DERIVED_TV_STATUS": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
    "derived": random_boolean(),
  },

  "AUDIO_SOURCE": {
            "ID": random_number(),
            "TS": current_timestamp(),
            "Type": random_number(),
    "source": random_audio_source(),
  },

    }

# Generate and print fake data
data = generate_device_data()
for key, value in data.items():
    print(f"{key}: {value}")

File: tests/fakeData.js
// const connectDB = require("./src/config/mongoDB");
const deviceSchema = require("../models/deviceData");
// const deviceAlertsSchema = require("./src/models/alertsModel");
const { faker } = require("@faker-js/faker");

const mongoose = require("mongoose");
require("dotenv").config();

const connectDB = async () => {
  try {
    await mongoose.connect(
      "mongodb+srv://ankurauti:ankurauti02@cluster0.ng3tq.mongodb.net/APM2?retryWrites=true&w=majority&appName=Cluster0"
    );
    console.log("MongoDB connected");
  } catch (err) {
    console.error(err.message);
    process.exit(1);
  }
};
connectDB();

const APM1Ids = [
  100001, 100002, 100003, 100004, 100005, 100006, 100007, 100008, 100009,
  100010,
];

const APM2Ids = [
  200001, 200002, 200003, 200004, 200005, 200006, 200007, 200008, 200009,
  200010,
];
const APM3Ids = [
  300001, 300002, 300003, 300004, 300005, 300006, 300007, 300008, 300009,
  300010,
];

//   const randomAPM1id = APM1Ids[Math.floor(Math.random() * APM1Ids.length)];
//   const randomAPM2id = APM1Ids[Math.floor(Math.random() * APM2Ids.length)];
//   const randomAPM3id = APM1Ids[Math.floor(Math.random() * APM3Ids.length)];

// Function to randomly select one of the APM IDs
function getRandomDeviceId() {
  // Array of all device arrays
  const allIds = [APM1Ids, APM2Ids, APM3Ids];

  // Randomly select one of the ID arrays
  const randomIdsArray = allIds[Math.floor(Math.random() * allIds.length)];

  // Randomly select an ID from the chosen array
  const randomDeviceId =
    randomIdsArray[Math.floor(Math.random() * randomIdsArray.length)];

  return randomDeviceId;
}

const dominicanRepublicLocations = [
  { lat: 18.5204, lon: 73.8567 }, // Pune
  { lat: 19.076, lon: 72.8777 }, // Mumbai
  { lat: 20.0074, lon: 73.7898 }, // Nashik
  { lat: 18.922, lon: 72.8347 }, // Thane
  { lat: 19.2183, lon: 73.0978 }, // Kalyan
  { lat: 18.9696, lon: 72.8205 }, // Dombivli
  { lat: 18.5074, lon: 73.8077 }, // Pimpri-Chinchwad
  { lat: 20.011, lon: 73.7602 }, // Nashik Road
  { lat: 19.9975, lon: 73.7898 }, // Deolali
  { lat: 19.845, lon: 72.9081 }, // Palghar
];

let lastDeviceId = 0;
let lastAlertId = 0;

function generateRandomData() {
  lastDeviceId++;

  const randomDominicanRepublicLocation =
    dominicanRepublicLocations[
      Math.floor(Math.random() * dominicanRepublicLocations.length)
    ];

  const alertTypes = ["Generated", "Pending", "Resolved"];
  const randomAlertType =
    alertTypes[Math.floor(Math.random() * alertTypes.length)];

  const HWversion = ["APM1", "APM2", "APM3"];
  const randomHWversion =
    HWversion[Math.floor(Math.random() * alertTypes.length)];

  return {
    DEVICE_ID: 100002,
    LOCATION: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 1,
      Cell_Info: {
        region: "India",
        lat: randomDominicanRepublicLocation.lat,
        lon: randomDominicanRepublicLocation.lon,
      },
      Installing: faker.datatype.boolean(),
    },
    GUEST_REGISTRATION: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 2,
      guest_id: faker.number.int({ min: 1, max: 6 }),
      registering: faker.datatype.boolean(),
      guest_age: faker.number.int({ min: 6, max: 80 }),
      guest_male: faker.datatype.boolean(),
    },
    MEMBER_GUEST_DECLARATION: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 3,
      member_keys: Array(12)
        .fill()
        .map(() => faker.datatype.boolean()),
      guests: Array(5)
        .fill()
        .map(() => faker.datatype.boolean()),
      confidence: 80,
    },
    CONFIGURATION: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 4,
      differential_mode: faker.datatype.boolean(),
      member_keys: Array(12)
        .fill()
        .map(() => faker.datatype.boolean()),
      guest_cancellation_time: 300,
      software_version: faker.system.semver(),
      hardware_version: randomHWversion,
      power_pcb_firmware_version: faker.system.semver(),
      remote_firmware_version: faker.system.semver(),
      audio_configuration: Array(5)
        .fill()
        .map(() => faker.datatype.boolean()),
      audience_day_start_time: 600,
      no_of_sessions: 8,
    },
    TAMPER_ALARM: {
      ID: lastDeviceId,
      AlertType: randomAlertType,
      Type: 5,
      TS: Date.now(),
      Meter_tamper: faker.datatype.boolean(),
      Tv_plug_tamper: faker.datatype.boolean(),
      Tamper_ts: Date.now(),
    },
    SOS_ALARM: {
      ID: lastDeviceId,
      AlertType: randomAlertType,
      Type: 6,
      TS: Date.now(),
      sos: faker.datatype.boolean(),
    },
    BATTERY_ALARM: {
      ID: lastDeviceId,
      AlertType: randomAlertType,
      Type: 7,
      TS: Date.now(),
      main_bat_fail: faker.datatype.boolean(),
      rtc_fail: faker.datatype.boolean(),
      rtc_bat_low: faker.datatype.boolean(),
    },
    METER_INSTALLATION: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 8,
      HHID: "125",
      Success: faker.datatype.boolean(),
    },
    VOLTAGE_STATS: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 9,
      high_rail_voltage: 3000,
      mid_rail_voltage: 1500,
      gsm_rail_voltage: 3300,
      rtc_battery_voltage: 3200,
      li_ion_battery_voltage: 3800,
      remote_battery_voltage: 2500,
    },
    TEMPERATURE_STATS: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 10,
      battery_temp: 25,
      arm_core_temp: 30,
      power_pcb_temp: 28,
      rtc_temp: 27,
    },
    NTP_SYNC: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 11,
      server: "time.google.com",
      system_time: Date.now(),
      success: faker.datatype.boolean(),
      error_code: 0,
      drift: 5,
      jumped: faker.datatype.boolean(),
    },
    AUDIENCE_SESSION_CLOSE: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 12,
      stop_time: 930,
      viewing_member_keys: Array(12)
        .fill()
        .map(() => faker.datatype.boolean()),
      viewing_guests: Array(5)
        .fill()
        .map(() => faker.datatype.boolean()),
      tv_on: faker.datatype.boolean(),
      last_watermark_id: 123,
      tv_event_ts: 456,
      last_watermark_id_ts: Date.now(),
      marked: 80,
    },
    NETWORK_LATCH: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 13,
      Ip_up: faker.datatype.boolean(),
      Sim: faker.number.int({ min: 1, max: 2 }),
    },
    REMOTE_PAIRING: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 14,
      remote_id: 123456789,
      success: faker.datatype.boolean(),
    },
    REMOTE_ACTIVITY: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 15,
      lock: faker.datatype.boolean(),
      orr: faker.datatype.boolean(),
      absent_key_press: faker.datatype.boolean(),
      drop: faker.datatype.boolean(),
    },
    SIM_ALERT: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 16,
      AlertType: randomAlertType,
      sim1_absent: faker.datatype.boolean(),
      sim1_changed: faker.datatype.boolean(),
      sim2_absent: faker.datatype.boolean(),
      sim2_changed: faker.datatype.boolean(),
    },
    SYSTEM_ALARM: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 17,
      name: "System Alarm",
      AlertType: randomAlertType,
      Details: {
        name: "System Error",
        error_code: faker.number.int({ min: 400, max: 599 }),
        message: faker.lorem.words({ min: 5, max: 10 }),
      },
    },
    SYSTEM_INFO: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 18,
      rpi_serial: faker.string.alphanumeric(10),
      pcb_serial: faker.string.alphanumeric(6),
      imei: faker.string.numeric(15),
      imsi_1: faker.string.numeric(15),
      imsi_2: faker.string.numeric(15),
      eeprom: faker.number.int(),
      wifi_serial: faker.string.alphanumeric(7),
      mac_serial: faker.string.alphanumeric(6),
      remote_serial: faker.number.int(),
    },
    CONFIG_UPDATE: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 19,
      key: "timeout",
      value: "60",
      old_value: "30",
    },
    ALIVE: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 20,
      state: faker.datatype.boolean(),
    },
    METER_OTA: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 21,
      previous: faker.system.semver(),
      update: faker.system.semver(),
      success: faker.datatype.boolean(),
    },
    BATTERY_VOLTAGE: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 22,
      Rtc: 2882,
      Meter: 4164,
    },
    BOOT: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 23,
      boot_ts: Date.now(),
      last_boot_ts: Date.now(),
      last_shutdown_ts: Date.now(),
      clean: faker.datatype.boolean(),
    },
    BOOT_V2: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 24,
      rtc_ts: Date.now(),
      ntp_ts: Date.now(),
      cell_ts: Date.now(),
      boot_ts: Date.now(),
      clean: faker.datatype.boolean(),
      time_approximated: faker.datatype.boolean(),
      events_ignored: faker.datatype.boolean(),
    },
    STB: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 25,
      state: faker.datatype.boolean(),
    },
    DERIVED_TV_STATUS: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 26,
      state: faker.datatype.boolean(),
    },
    AUDIO_SOURCE: {
      ID: lastDeviceId,
      TS: Date.now(),
      Type: 27,
      Status: "line_in",
    },
  };
}

async function saveDataToMongo(inputJSON, schema) {
  try {
    const result = await schema.insertMany(inputJSON);
    if (result.length > 0) {
      return {
        statusCode: 200,
        message: "Document inserted successfully",
      };
    } else {
      throw new Error("Document not inserted");
    }
  } catch (error) {
    throw new Error(error.message);
  }
}

function sendDataToMongoEvery5min() {
  setInterval(() => {
    const dummyData = generateRandomData();
    saveDataToMongo([dummyData], deviceSchema)
      .then((result) => {
        console.log(result.message);
      })
      .catch((error) => {
        console.error(error.message);
      });
  }, 5000); // 5 seconds in milliseconds
}

sendDataToMongoEvery5min();
